File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/context/EditContextService.kt
/** Edit Context Service — source + callers + tests + git for editing. KSA-171. */
package com.codeintel.context

import com.codeintel.context.models.*
import com.codeintel.graph.CallGraphService
import com.codeintel.graph.SymbolResolver
import com.codeintel.graph.TestDetector
import java.io.File
import java.sql.Connection

class EditContextService(
    private val conn: Connection,
    private val resolver: SymbolResolver,
    private val callGraph: CallGraphService,
    private val testDetector: TestDetector,
    private val workspace: String,
) {
    private val gitService = GitService(workspace)
    private val budgetMgr = TokenBudgetManager(4000)

    /** Get full edit context for a symbol. */
    fun getContext(params: EditContextParams): EditContextResult {
        val startTime = System.currentTimeMillis()

        val symbol = resolveSymbolInput(params.symbol)
            ?: return notFound(params.symbol, params.tokenBudget, startTime)

        val source = readSymbolSource(symbol)
        val signature = getSignature(symbol)

        val callers = if (params.includeCallers) getCallerContext(symbol, params.callerDepth) else null
        val tests = if (params.includeTests) getTestContext(symbol) else null
        val gitHistory = if (params.includeGit) getGitContext(symbol) else null
        val siblings = getSiblingContext(symbol)

        // Assemble within budget
        val sections = mutableMapOf<String, Pair<String, Int>>()
        sections["source"] = source to 1
        callers?.let { sections["callers"] = serializeCallers(it) to 2 }
        tests?.let { sections["tests"] = serializeTests(it) to 3 }
        gitHistory?.let { sections["git_history"] = serializeGit(it) to 5 }
        siblings?.let { sections["siblings"] = serializeSiblings(it) to 6 }

        val assembled = budgetMgr.assemble(sections, params.tokenBudget)
        val elapsed = System.currentTimeMillis() - startTime

        return EditContextResult(
            symbol = symbol.name,
            file = symbol.filePath,
            line = symbol.line,
            kind = symbol.kind,
            source = assembled.result["source"] ?: source,
            signature = signature,
            callers = if ("callers" in assembled.result) callers else null,
            tests = if ("tests" in assembled.result) tests else null,
            gitHistory = if ("git_history" in assembled.result) gitHistory else null,
            siblings = if ("siblings" in assembled.result) siblings else null,
            metadata = EditMetadata(
                tokenCount = assembled.tokenCount,
                tokenBudget = params.tokenBudget,
                sectionsIncluded = assembled.included,
                sectionsExcluded = assembled.excluded,
                queryTimeMs = elapsed,
            ),
        )
    }

    private fun resolveSymbolInput(input: String): ResolvedSymbolFull? {
        // Try file:line format
        if (":" in input && input.matches(Regex(".*:\\d+$"))) {
            val colonIdx = input.lastIndexOf(':')
            val filePart = input.substring(0, colonIdx)
            val line = input.substring(colonIdx + 1).toInt()
            return findSymbolAtLine(filePart, line)
        }

        val resolved = resolver.resolve(input)
        if (resolved.isEmpty()) return null

        val sym = resolved[0]
        val stmt = conn.prepareStatement("SELECT end_line, signature FROM symbols WHERE id = ?")
        stmt.setInt(1, sym.id)
        val rs = stmt.executeQuery()
        val endLine = if (rs.next()) rs.getInt(1).takeIf { !rs.wasNull() } else null
        val sig = rs.getString(2)

        return ResolvedSymbolFull(sym.id, sym.name, sym.kind, sym.filePath, sym.line, sym.parentSymbolId, endLine, sig)
    }

    private fun findSymbolAtLine(filePart: String, line: Int): ResolvedSymbolFull? {
        val stmt = conn.prepareStatement("""
            SELECT s.id, s.name, s.kind, f.relative_path, s.start_line, s.end_line, s.signature, s.parent_symbol
            FROM symbols s JOIN files f ON s.file_id = f.id
            WHERE f.relative_path LIKE ? AND s.start_line <= ? AND s.end_line >= ?
            ORDER BY (s.end_line - s.start_line) ASC LIMIT 1
        """.trimIndent())
        stmt.setString(1, "%$filePart")
        stmt.setInt(2, line)
        stmt.setInt(3, line)
        val rs = stmt.executeQuery()
        if (!rs.next()) return null
        val parentStr = rs.getString(8)
        return ResolvedSymbolFull(
            rs.getInt(1), rs.getString(2), rs.getString(3), rs.getString(4),
            rs.getInt(5), parentStr?.toIntOrNull(), rs.getInt(6).takeIf { !rs.wasNull() }, rs.getString(7),
        )
    }

    private fun readSymbolSource(symbol: ResolvedSymbolFull): String {
        return try {
            val lines = File(workspace, symbol.filePath).readLines()
            val start = symbol.line - 1
            val end = symbol.endLine ?: (start + 50)
            lines.subList(start.coerceAtLeast(0), end.coerceAtMost(lines.size)).joinToString("\n")
        } catch (_: Exception) { "" }
    }

    private fun getSignature(symbol: ResolvedSymbolFull): String? {
        if (symbol.signature != null) return symbol.signature
        val stmt = conn.prepareStatement("SELECT signature FROM symbols WHERE id = ?")
        stmt.setInt(1, symbol.id)
        val rs = stmt.executeQuery()
        return if (rs.next()) rs.getString(1) else null
    }

    private fun getCallerContext(symbol: ResolvedSymbolFull, depth: Int): List<CallerContext>? {
        val result = callGraph.findCallers(symbol.name, depth, 10)
        if (result.results.isEmpty()) return null
        return result.results.map { caller ->
            val ctx = getLineContext(caller.filePath, caller.callSiteLine, 2)
            CallerContext(
                symbol = caller.qualifiedName.ifEmpty { caller.symbol },
                file = caller.filePath, line = caller.callSiteLine, context = ctx,
            )
        }
    }

    private fun getLineContext(filePath: String, line: Int, surrounding: Int): String {
        return try {
            val lines = File(workspace, filePath).readLines()
            val start = maxOf(0, line - 1 - surrounding)
            val end = minOf(lines.size, line + surrounding)
            lines.subList(start, end).joinToString("\n")
        } catch (_: Exception) { "" }
    }

    private fun getTestContext(symbol: ResolvedSymbolFull): List<TestContext>? {
        val resolvedSym = com.codeintel.graph.models.ResolvedSymbol(
            id = symbol.id, name = symbol.name, kind = symbol.kind,
            filePath = symbol.filePath, line = symbol.line, parentSymbolId = symbol.parentSymbolId,
        )
        val testFiles = testDetector.findRelatedTests(listOf(resolvedSym), emptyList())
        if (testFiles.isEmpty()) return null
        val results = mutableListOf<TestContext>()
        for (tf in testFiles.take(3)) {
            try {
                val content = File(workspace, tf.file).readText()
                val blocks = extractTestBlocks(content, symbol.name)
                for (block in blocks.take(2)) {
                    results.add(TestContext(file = tf.file, testName = block.first, source = block.second))
                }
            } catch (_: Exception) { continue }
        }
        return results.ifEmpty { null }
    }

    private fun extractTestBlocks(content: String, symbolName: String): List<Pair<String, String>> {
        val blocks = mutableListOf<Pair<String, String>>()
        val lines = content.lines()
        val pattern = Regex("""(?:it|test|describe|@Test)\s*[\('"](.*?)['"]""")

        for ((i, line) in lines.withIndex()) {
            val match = pattern.find(line) ?: continue
            val window = lines.subList(i, minOf(i + 10, lines.size)).joinToString("\n")
            if (symbolName in line || symbolName in window) {
                val name = match.groupValues[1]
                val end = minOf(i + 15, lines.size)
                val source = lines.subList(i, end).joinToString("\n")
                blocks.add(name to source)
            }
        }
        return blocks
    }

    private fun getGitContext(symbol: ResolvedSymbolFull): List<GitCommit>? {
        val commits = gitService.getFileHistory(symbol.filePath, 5)
        return commits.ifEmpty { null }
    }

    private fun getSiblingContext(symbol: ResolvedSymbolFull): List<SiblingContext>? {
        val (sql, params) = if (symbol.parentSymbolId != null) {
            "SELECT name, kind, signature, start_line FROM symbols WHERE parent_symbol = ? AND id != ? ORDER BY start_line" to
                listOf(symbol.parentSymbolId.toString(), symbol.id)
        } else {
            """SELECT s.name, s.kind, s.signature, s.start_line
               FROM symbols s JOIN files f ON s.file_id = f.id
               WHERE f.relative_path = ? AND s.parent_symbol IS NULL AND s.id != ?
               ORDER BY s.start_line""" to listOf(symbol.filePath, symbol.id)
        }
        val stmt = conn.prepareStatement(sql)
        stmt.setString(1, params[0].toString())
        stmt.setInt(2, params[1] as Int)
        val rs = stmt.executeQuery()
        val siblings = mutableListOf<SiblingContext>()
        while (rs.next()) siblings.add(SiblingContext(rs.getString(1), rs.getString(2), rs.getString(3), rs.getInt(4)))
        return siblings.ifEmpty { null }
    }

    private fun serializeCallers(callers: List<CallerContext>) = callers.joinToString("\n") { "${it.symbol}|${it.file}:${it.line}" }
    private fun serializeTests(tests: List<TestContext>) = tests.joinToString("\n") { "${it.testName}|${it.file}" }
    private fun serializeGit(commits: List<GitCommit>) = commits.joinToString("\n") { "${it.hash} ${it.message}" }
    private fun serializeSiblings(siblings: List<SiblingContext>) = siblings.joinToString("\n") { "${it.name}|${it.kind}|${it.line}" }

    private fun notFound(symbol: String, budget: Int, startTime: Long): EditContextResult {
        return EditContextResult(
            symbol = symbol, file = "", line = 0, kind = "unknown", source = "",
            metadata = EditMetadata(tokenBudget = budget, sectionsExcluded = listOf("error: symbol not found"),
                queryTimeMs = System.currentTimeMillis() - startTime),
        )
    }
}

internal data class ResolvedSymbolFull(
    val id: Int, val name: String, val kind: String, val filePath: String,
    val line: Int, val parentSymbolId: Int?, val endLine: Int?, val signature: String?,
)

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/context/edit_context_service.py
"""Edit Context Service — source + callers + tests + git for editing. KSA-171."""

from __future__ import annotations

import os
import re
import sqlite3
import time
from typing import Any

from ..graph.call_graph_service import CallGraphService
from ..graph.symbol_resolver import SymbolResolver
from ..graph.test_detector import TestDetector
from .git_service import GitService
from .token_budget_manager import TokenBudgetManager
from .types import (
    CallerContext, EditContextParams, EditContextResult,
    GitCommit, SiblingContext, TestContext,
)


class EditContextService:
    """Gathers everything needed before modifying a symbol."""

    def __init__(
        self,
        conn: sqlite3.Connection,
        resolver: SymbolResolver,
        call_graph: CallGraphService,
        test_detector: TestDetector,
        workspace: str,
    ) -> None:
        self._conn = conn
        self._resolver = resolver
        self._call_graph = call_graph
        self._test_detector = test_detector
        self._git_service = GitService(workspace)
        self._budget_mgr = TokenBudgetManager(4000)
        self._workspace = workspace

    def get_context(self, params: EditContextParams) -> EditContextResult:
        """Get full edit context for a symbol."""
        start_time = time.time()

        # 1. Resolve symbol
        symbol = self._resolve_symbol_input(params.symbol)
        if not symbol:
            return self._not_found(params.symbol, params.token_budget, start_time)

        # 2. Read source (always included)
        source = self._read_symbol_source(symbol)
        signature = self._get_signature(symbol)

        # 3. Gather sections
        callers = self._get_caller_context(symbol, params.caller_depth) if params.include_callers else None
        tests = self._get_test_context(symbol) if params.include_tests else None
        git_history = self._get_git_context(symbol) if params.include_git else None
        siblings = self._get_sibling_context(symbol)

        # 4. Assemble within token budget
        sections: dict[str, dict[str, Any]] = {
            "source": {"content": source, "priority": 1},
        }
        if callers:
            sections["callers"] = {"content": callers, "priority": 2}
        if tests:
            sections["tests"] = {"content": tests, "priority": 3}
        if git_history:
            sections["git_history"] = {"content": git_history, "priority": 5}
        if siblings:
            sections["siblings"] = {"content": siblings, "priority": 6}

        assembled = self._budget_mgr.assemble(sections, params.token_budget)

        elapsed_ms = int((time.time() - start_time) * 1000)
        result = EditContextResult(
            symbol=symbol["name"],
            file=symbol["file_path"],
            line=symbol["line"],
            kind=symbol["kind"],
            source=assembled["result"].get("source", source),
            signature=signature,
            metadata={
                "tokenCount": assembled["token_count"],
                "tokenBudget": params.token_budget,
                "sectionsIncluded": assembled["included"],
                "sectionsExcluded": assembled["excluded"],
                "queryTimeMs": elapsed_ms,
            },
        )

        if "callers" in assembled["result"]:
            result.callers = assembled["result"]["callers"]
        if "tests" in assembled["result"]:
            result.tests = assembled["result"]["tests"]
        if "git_history" in assembled["result"]:
            result.git_history = assembled["result"]["git_history"]
        if "siblings" in assembled["result"]:
            result.siblings = assembled["result"]["siblings"]

        return result

    def _resolve_symbol_input(self, input_str: str) -> dict[str, Any] | None:
        # Try file:line format
        if ":" in input_str and re.search(r":\d+$", input_str):
            colon_idx = input_str.rfind(":")
            file_part = input_str[:colon_idx]
            line = int(input_str[colon_idx + 1:])
            return self._find_symbol_at_line(file_part, line)

        # Standard resolution
        resolved = self._resolver.resolve(input_str)
        if not resolved:
            return None

        sym = resolved[0]
        cur = self._conn.execute(
            "SELECT end_line, signature FROM symbols WHERE id = ?", (sym.id,)
        )
        row = cur.fetchone()
        return {
            "id": sym.id,
            "name": sym.name,
            "kind": sym.kind,
            "file_path": sym.file_path,
            "line": sym.line,
            "end_line": row[0] if row else None,
            "signature": row[1] if row else None,
            "parent_symbol_id": sym.parent_symbol_id,
        }

    def _find_symbol_at_line(self, file_part: str, line: int) -> dict[str, Any] | None:
        cur = self._conn.execute(
            """SELECT s.id, s.name, s.kind, f.relative_path, s.start_line,
                      s.end_line, s.signature, s.parent_symbol
               FROM symbols s JOIN files f ON s.file_id = f.id
               WHERE f.relative_path LIKE ? AND s.start_line <= ? AND s.end_line >= ?
               ORDER BY (s.end_line - s.start_line) ASC LIMIT 1""",
            (f"%{file_part}", line, line),
        )
        row = cur.fetchone()
        if not row:
            return None
        parent_id = int(row[7]) if row[7] and str(row[7]).isdigit() else None
        return {
            "id": row[0], "name": row[1], "kind": row[2], "file_path": row[3],
            "line": row[4], "end_line": row[5], "signature": row[6],
            "parent_symbol_id": parent_id,
        }

    def _read_symbol_source(self, symbol: dict[str, Any]) -> str:
        try:
            full_path = os.path.join(self._workspace, symbol["file_path"])
            with open(full_path, encoding="utf-8") as f:
                lines = f.readlines()
            start = symbol["line"] - 1
            end = symbol.get("end_line") or start + 50
            return "".join(lines[start:end])
        except (OSError, IndexError):
            return ""

    def _get_signature(self, symbol: dict[str, Any]) -> str | None:
        if symbol.get("signature"):
            return symbol["signature"]
        cur = self._conn.execute("SELECT signature FROM symbols WHERE id = ?", (symbol["id"],))
        row = cur.fetchone()
        return row[0] if row else None

    def _get_caller_context(self, symbol: dict[str, Any], depth: int) -> list[CallerContext] | None:
        result = self._call_graph.find_callers(symbol["name"], depth, 10)
        if not result.results:
            return None
        callers: list[CallerContext] = []
        for caller in result.results:
            ctx = self._get_line_context(caller.file_path, caller.call_site_line, 2)
            callers.append(CallerContext(
                symbol=caller.qualified_name or caller.symbol,
                file=caller.file_path,
                line=caller.call_site_line,
                context=ctx,
            ))
        return callers

    def _get_line_context(self, file_path: str, line: int, surrounding: int) -> str:
        try:
            full_path = os.path.join(self._workspace, file_path)
            with open(full_path, encoding="utf-8") as f:
                lines = f.readlines()
            start = max(0, line - 1 - surrounding)
            end = min(len(lines), line + surrounding)
            return "".join(lines[start:end])
        except (OSError, IndexError):
            return ""

    def _get_test_context(self, symbol: dict[str, Any]) -> list[TestContext] | None:
        from ..graph.models import ResolvedSymbol
        sym = ResolvedSymbol(
            id=symbol["id"], name=symbol["name"], kind=symbol["kind"],
            file_path=symbol["file_path"], line=symbol["line"],
            parent_symbol_id=symbol.get("parent_symbol_id"),
        )
        test_files = self._test_detector.find_related_tests([sym], [])
        results: list[TestContext] = []
        for tf in test_files[:3]:
            try:
                full_path = os.path.join(self._workspace, tf.file)
                with open(full_path, encoding="utf-8") as f:
                    content = f.read()
                blocks = self._extract_test_blocks(content, symbol["name"])
                for block in blocks[:2]:
                    results.append(TestContext(
                        file=tf.file, test_name=block["name"], source=block["source"],
                    ))
            except OSError:
                continue
        return results if results else None

    def _extract_test_blocks(self, content: str, symbol_name: str) -> list[dict[str, str]]:
        blocks: list[dict[str, str]] = []
        lines = content.splitlines()
        test_pattern = re.compile(r"""(?:it|test|describe|def test_)\s*[\('"](.*?)['"]""")

        for i, line in enumerate(lines):
            match = test_pattern.search(line)
            if match:
                window = "\n".join(lines[i:i + 10])
                if symbol_name in line or symbol_name in window:
                    name = match.group(1)
                    end = min(i + 15, len(lines))
                    source = "\n".join(lines[i:end])
                    blocks.append({"name": name, "source": source})
        return blocks

    def _get_git_context(self, symbol: dict[str, Any]) -> list[GitCommit] | None:
        commits = self._git_service.get_file_history(symbol["file_path"], 5)
        return commits if commits else None

    def _get_sibling_context(self, symbol: dict[str, Any]) -> list[SiblingContext] | None:
        if symbol.get("parent_symbol_id"):
            cur = self._conn.execute(
                "SELECT name, kind, signature, start_line FROM symbols WHERE parent_symbol = ? AND id != ? ORDER BY start_line",
                (str(symbol["parent_symbol_id"]), symbol["id"]),
            )
        else:
            cur = self._conn.execute(
                """SELECT s.name, s.kind, s.signature, s.start_line
                   FROM symbols s JOIN files f ON s.file_id = f.id
                   WHERE f.relative_path = ? AND s.parent_symbol IS NULL AND s.id != ?
                   ORDER BY s.start_line""",
                (symbol["file_path"], symbol["id"]),
            )
        rows = cur.fetchall()
        if not rows:
            return None
        return [SiblingContext(name=r[0], kind=r[1], signature=r[2], line=r[3]) for r in rows]

    def _not_found(self, symbol: str, budget: int, start_time: float) -> EditContextResult:
        elapsed_ms = int((time.time() - start_time) * 1000)
        return EditContextResult(
            symbol=symbol, file="", line=0, kind="unknown",
            source="", signature=None,
            metadata={
                "tokenCount": 0, "tokenBudget": budget,
                "sectionsIncluded": [], "sectionsExcluded": ["error: symbol not found"],
                "queryTimeMs": elapsed_ms,
            },
        )
